### ISSUE TITLE
UI::TextEdit caret position missing and not scrolling text to the left

### ISSUE THREAD

**Issue by adenovan** [hrydgard/ppsspp#10026]
Before exceed bounds.w
![screenshot 435](https://user-images.githubusercontent.com/5246143/31853985-aef37902-b6c4-11e7-86ab-9c0ebb074864.png)

After exceed bounds.w 
![screenshot 436](https://user-images.githubusercontent.com/5246143/31853987-bdbd2c1c-b6c4-11e7-8b83-7f0285d5878f.png)

looks like https://github.com/hrydgard/ppsspp/blob/master/ext/native/ui/view.cpp#L803 scroll code not working?

**Comment by hrydgard** [comment]
Hm, I first thought the TextEdit was cut off by the surrounding container but it does look like the width is right. Strange... seems like a bug, yes.

**Comment by adenovan** [comment]
the variable textX seems did not move the TextEdit text_ i already uncomment it and its does nothing, 
i found a way to simulate scroll left by creating temporary string display variable on this line https://github.com/hrydgard/ppsspp/blob/master/ext/native/ui/view.cpp#L794 . but its hard to positioning the caret position after exceeding the bounds width.

**Comment by adenovan** [comment]
a bit slow but do the job well. neex new private maxTextDisplay_ on view.h @hrydgard 
```
void TextEdit::Draw(UIContext &dc) {
	dc.PushScissor(bounds_);
	dc.SetFontStyle(dc.theme->uiFont);
	dc.FillRect(HasFocus() ? UI::Drawable(0x80000000) : UI::Drawable(0x30000000), bounds_);

	uint32_t textColor = hasTextColor_ ? textColor_ : dc.theme->infoStyle.fgColor;
	float textX = bounds_.x;
	float w, h;

	Bounds textBounds = bounds_;
	textBounds.x = textX;
	std::string draw;
	int subcount = 0;

	if (text_.empty()) {
		if (placeholderText_.size()) {
			uint32_t c = textColor & 0x50FFFFFF;
			dc.DrawTextRect(placeholderText_.c_str(), bounds_, c, ALIGN_CENTER);
		}
		maxTextDisplay_ = 0;
	} else {
		// scroll left if exceed width
		if (maxTextDisplay_ != 0 && caret_ >= maxTextDisplay_) {
			draw = text_.substr(text_.length() - maxTextDisplay_, text_.length());
			dc.DrawTextRect(draw.c_str(), textBounds, textColor, ALIGN_VCENTER | ALIGN_LEFT);
			subcount += 1;
		}
		else {
			//reset the scroll
			maxTextDisplay_ = 0;
			subcount = 0;
			draw = text_.substr(0,text_.length());
			dc.DrawTextRect(text_.c_str(), textBounds, textColor, ALIGN_VCENTER | ALIGN_LEFT);
		}

	}

	if (HasFocus()) {
	
		if (maxTextDisplay_ != 0 && caret_ >= maxTextDisplay_) {
			// measure by minus substring count
			dc.MeasureTextCount(dc.theme->uiFont, 1.0f, 1.0f, text_.c_str(), caret_ - subcount, &w, &h, ALIGN_VCENTER | ALIGN_LEFT);
		}
		else {
			// Hack to find the caret position. Might want to find a better way..
			dc.MeasureTextCount(dc.theme->uiFont, 1.0f, 1.0f, text_.c_str(), caret_, &w, &h, ALIGN_VCENTER | ALIGN_LEFT);
		}
		float caretX = w;
		caretX += textX;

		if (caretX >= bounds_.w) {
			// Scroll text to the left if the caret won't fit. Not ideal but looks better than not scrolling it.
			//textX -= caretX - bounds_.w;

			// store the caret max position
			if (maxTextDisplay_ == 0) {
				maxTextDisplay_ = caret_;
			}

			// decrement the caret position when exceed bounds
			caretX-= textX;
		}
		dc.FillRect(UI::Drawable(textColor), Bounds(caretX - 1, bounds_.y + 2, 3, bounds_.h - 4));
	}
	dc.PopScissor();
}
```

**Comment by adenovan** [comment]
@hrydgard the code is updated and its works. did you accept we scroll to the left using substring? if yes should i do separate pull request or combine it with the chat pull request #9090 ?

im working on chat gui right now but my chat server has different implementation with the proadhocserver, the server is on separate project , it will be a large pull request if  want to be merged with ppsspp, once the server code is complete i will upload it to github and let you know @hrydgard it also will be used in the implementation of tunneling game packet by depending on a single server with public ip. here is some screenshot 

![screenshot 450](https://user-images.githubusercontent.com/5246143/31920533-5966c3e8-b89c-11e7-8e2d-f673fed323f5.png)
![screenshot 451](https://user-images.githubusercontent.com/5246143/31920534-599e5d44-b89c-11e7-8f5f-97a2ea39b24f.png)
![screenshot 452](https://user-images.githubusercontent.com/5246143/31920535-59d821c8-b89c-11e7-85a3-3b5445e0579b.png)
![screenshot 453](https://user-images.githubusercontent.com/5246143/31920536-5a0f6a2a-b89c-11e7-92a6-68d5a21e678b.png)
![screenshot 454](https://user-images.githubusercontent.com/5246143/31920537-5a42ec4c-b89c-11e7-9e86-2b70aa326267.png)
![screenshot 455](https://user-images.githubusercontent.com/5246143/31920538-5a77482a-b89c-11e7-9f1b-8188b76a817d.png)
![screenshot 456](https://user-images.githubusercontent.com/5246143/31920539-5aad7594-b89c-11e7-8d9e-106df82bfd27.png)
![screenshot 457](https://user-images.githubusercontent.com/5246143/31920540-5ae9fcbc-b89c-11e7-93ed-11e7d98f42e7.png)
![screenshot 458](https://user-images.githubusercontent.com/5246143/31920541-5b3d28ce-b89c-11e7-87ea-3c096e4470c0.png)
![screenshot 459](https://user-images.githubusercontent.com/5246143/31920543-5b9db37e-b89c-11e7-8838-bdf2ca722d85.png)
![screenshot 460](https://user-images.githubusercontent.com/5246143/31920545-5bfea9fe-b89c-11e7-99c7-c631a1673b5b.png)
![screenshot 461](https://user-images.githubusercontent.com/5246143/31920549-5c6544e8-b89c-11e7-9a85-f1d57a7e3998.png)
![screenshot 462](https://user-images.githubusercontent.com/5246143/31920550-5cd8c148-b89c-11e7-976e-d47a3e3f30e8.png)

i also want to implement the chat On Screen Message to quickview other player incoming chat. and i need a chat logo too and use it to substitute the button.  for the OSM should i modify the current OSM or create another chat OSM?

**Comment by hrydgard** [comment]
Sorry for the delay, thanks, I'll take a look at your change shortly. Your chat stuff is starting to look good.

If the chat OSM stuff is intended to look quite a bit different and be in a different position on the screen, you might want to add a separate one. Otherwise it might be better to have them the same so they don't overlap each other...

**Comment by hrydgard** [comment]
Got a change coming up soon with a slightly different fix for the caret scrolling.

**Comment by hrydgard** [comment]
There. Now you can rebase your big chat pull request on master :)

**Comment by adenovan** [comment]
ok thanks for the reply and the fix. i will start to add separate chat OSM so it will looks much better if the chat screen on close state
